emoji_picker: Extend emoji picker to show the name of emojis while navigating.

HarshitOnGitHub opened the pull request:
![anim](https://user-images.githubusercontent.com/16687990/29753937-23b40fa6-8b99-11e7-9bf8-9a8899a1b93f.gif)
This PR is based off of my another PR #6268 which is a per-requisite for this PR. Currently the realm emoji images are not properly aligned, PR #6089 will fix it.
Only issue remaining to fix:
How to properly wrap very long emoji names like `raised_hand_with_splayed_fingers`

![image](https://user-images.githubusercontent.com/16687990/29753978-edf7246a-8b99-11e7-90a0-cbf7b083f0bf.png)

Fixes: #6110 and #6111.

timabbott commented:
For too long names, I'd just do overflow=elipses to handle too-long names, with a mouse-over to see the full name in a `title` attribute or something.  Do we think the second row with the name containing `:`s is actually useful, or should we just have the emoji name in the bottom?

HarshitOnGitHub commented:
@timabbott I have updated this(currently rebased on the top of my PR #6250). This should somewhat ease the manual testing of PR #6250.

>Do we think the second row with the name containing :s is actually useful, or should we just have the emoji name in the bottom?

Currently there are not a very large number of emojis which have aliases and displaying the name of the emoji in the same format again was looking a bit awkward so I changed it to display the syntax we use for inserting emojis, since we are currently not displaying it anywhere in the UI. Plus my argument also got the support of the fact that it is how slack and WDT Emoji Bundle(https://ned.im/wdt-emoji-bundle/) do it.

HarshitOnGitHub commented:
Just a note, this not ready to be merged yet.

timabbott commented:
@HarshitOnGitHub how close is this to being mergable?  If it makes sense, I can start reviewing and merging preparatory commits.

HarshitOnGitHub commented:
@timabbott This PR needs a little bit more work, I will be able to finish it soon! It is currently based on PR #6268 so it will be good if you can give PR #6268 a review. :)

timabbott commented:
OK, that's merged; so you can rebase this (do you know why Travis failed?)

HarshitOnGitHub commented:
Not sure but seems unrelated to changes, will debug and let you know. :)

timabbott commented:
Oh, that's the casper test that's been flaking a lot.  We should probably disable it soon since it's proven hard to debug.  Feel free to ignore.

HarshitOnGitHub commented:
@timabbott I have updated this and it is ready for a review. This PR adds a space at the bottom of the emoji picker(that I refer to as emoji showcase) which displays some info about the currently focused emoji while navigating in the emoji picker by the arrow keys. For now I have kept back the commit that updated emoji showcase on mouse hover because it causing tracebacks for some emojis.
(Manually tested on both chrome as well as firefox on ubuntu).

timabbott commented:
I played with this, and it works pretty well.  There's 2 things that seem not great:
* It's kinda weird that it shows the Zulip logo by default when nothing is selected.
* I sorta would be happier if the showcase area just showed the emoji name with spaces (e.g. "clouds with lightning", without the `:` syntax.  The `:` syntax is easy to figure out if you need it (since that's what you get when you click the emoji in the compose emoji picker) and I feel like it clutters the picker.

HarshitOnGitHub commented:
>It's kinda weird that it shows the Zulip logo by default when nothing is selected.

@timabbott @rishig Any suggestions for what should be displayed when nothing is selected?

HarshitOnGitHub commented:
@timabbott Will this go in 1.7 or after that?

timabbott commented:
I'd take it for 1.7 if it's ready in the next day, since it is a nice user-facing improvement.

rishig commented:
It's worth noting we're about to change the default names, but I can try to get that in quickly if we want it for 1.7.

HarshitOnGitHub commented:
>
>*    It's kinda weird that it shows the Zulip logo by default when nothing is selected.
>*   I sorta would be happier if the showcase area just showed the emoji name with spaces (e.g. >"clouds with lightning", without the : syntax. The : syntax is easy to figure out if you need it (since >that's what you get when you click the emoji in the compose emoji picker) and I feel like it clutters the >picker.

@timabbott Second issue is easy to fix, for the first issue if you can suggest what to display when nothing is selected I can get it ready now.

timabbott commented:
Let's just have that space be blank initially.  And we can fill it in with something like a greyed-out placeholder-feeling smiley emoji in a follow-up project if we feel it looks wrong with it blank.

How does that sound?

HarshitOnGitHub commented:
@timabbott Done!

timabbott commented:
@HarshitOnGitHub this looks pretty good.  I have two changes to request:
(1) What I'd meant with the text in the showcase is that I think we should ONLY display the name of the highlighted emoji (e.g. "thumbs up") in bold; no need for 2 lines of text at all (I think it's basically a bug that Slack's emoji picker shows 2 duplicate lines in its showcase feature).  
(2) We'll probably want both the selected emoji and the showcase to update as one moves the mouse around over the emoji.  I imagine that it's own project, so we should make it a follow-up commit; can you open an issue and plan to work on it?  

@rishig do you agree?

rishig commented:
yup, agree with both. Also there is no need to display the alternate names anywhere; always having 1 line is fine.

HarshitOnGitHub commented:
@timabbott Updated to address your and rishi's comments.

timabbott commented:
This looks great, with the exceptions that:
* We should rebase
* It'd be really good to make this appear on mouseover
* The handling of too-long names isn't great.  Truncating is probably OK (you can get the name with the title mouseover in the picker, but we might just want to get rid of the `title` attribute once we have the showcase).  Long-term we may need to just make the picker wider, but I wonder if we could make 2 lines fit OK for things that wrap?  
![image](https://user-images.githubusercontent.com/2746074/31035410-b36f8304-a51c-11e7-9102-52dec9b42f75.png)

HarshitOnGitHub commented:
@timabbott I was about to push all the requested changes just before you commented but then got to fix some node tests.
> Truncating is probably OK (you can get the name with the title mouseover in the picker, but we might just want to get rid of the title attribute once we have the showcase).

I have truncated the names, removed the title attribute from the emojis and added title to the truncated name in emoji showcase itself.

Can you try out this now? :)

timabbott commented:
This feels pretty great.  Time to review the code :)

timabbott commented:
Nice, merged, thanks @HarshitOnGitHub!  I'll deploy this to chat.zulip.org soon.

The one follow-up thing we should either fix or open an issue for is that I think we should have the keyboard cursor move as you move your mouse to mark the emoji you currently would select.  And then if you switch back and forth between keyboard and mouse, the keyboard should start with where the mouse left off.

rishig commented:
For the long names: Long names are bad anyway, and hopefully there will be fewer after the renaming (I can also try to explicitly keep names short).

timabbott commented:
From my reading, I think some of these will be hard to make shorter without losing meaning, so I don't think we should try hard to remove long names.  Things like "stuck out tongue closed eyes" or "man and woman holding hands" are hard to shorten.  I'd range make it bigger if we feel change is needed.

rishig commented:
hm yeah, I agree some will be unavoidable. Another option is to make the font smaller if the name is too long.

timabbott commented:
Right.  I think the most likely solution is that we may want to make the picker 25% larger in general, and that would automatically resolve this.

timabbott commented:
Let's open an issue for improving our handling of long names in the emoji picker, with a screenshot of the current behavior, so we don't forget about it.
